File: src/Security/PassphraseManager.cs
namespace Pyxelze;

internal static class PassphraseManager
{
    private static readonly object _lock = new();
    private static string? _cachedPassphrase;

    public static string? CachedPassphrase
    {
        get { lock (_lock) return _cachedPassphrase; }
    }

    public static void Save(string pass)
    {
        lock (_lock) _cachedPassphrase = pass;
        Logger.Log("PassphraseManager: cached in memory");
    }

    public static void Clear()
    {
        lock (_lock) _cachedPassphrase = null;
        Logger.Log("PassphraseManager: cleared");
    }

    public static bool NeedsPassphrase(string stdout, string stderr)
    {
        return ContainsPassphraseIndicator(stdout) || ContainsPassphraseIndicator(stderr);
    }

    public static bool IsDecryptionFailure(string stdout, string stderr)
    {
        return Contains(stdout, "AES decryption failed") || Contains(stderr, "AES decryption failed")
            || Contains(stdout, "Encrypted payload") || Contains(stderr, "Encrypted payload");
    }

    private static bool ContainsPassphraseIndicator(string? text)
    {
        if (string.IsNullOrEmpty(text)) return false;
        return text.Contains("Passphrase required") || text.Contains("Encrypted payload") || text.Contains("AES decryption failed");
    }

    private static bool Contains(string? text, string value) =>
        text?.Contains(value) == true;

    public static string? PromptForPassphrase(string? errorMsg = null)
    {
        return PassphrasePrompt.Prompt("Passphrase requise", "Ce fichier est chiffré. Entrez la passphrase :", errorMsg);
    }

    public static string BuildPassphraseArg(string passphrase)
    {
        var escaped = passphrase.Replace("\"", "\\\"");
        return $"--passphrase \"{escaped}\"";
    }
}

File: src/security/PassphraseManager.cpp
#include "PassphraseManager.h"
#include <QMutex>

static QMutex s_mutex;
static QString s_cached;

static bool contains(const QString& text, const QString& needle) {
    return !text.isEmpty() && text.contains(needle, Qt::CaseInsensitive);
}

QString PassphraseManager::cachedPassphrase() {
    QMutexLocker lock(&s_mutex);
    return s_cached;
}

void PassphraseManager::save(const QString& pass) {
    QMutexLocker lock(&s_mutex);
    s_cached = pass;
}

void PassphraseManager::clear() {
    QMutexLocker lock(&s_mutex);
    s_cached.clear();
}

bool PassphraseManager::needsPassphrase(const QString& stdOut, const QString& stdErr) {
    return contains(stdOut, QStringLiteral("Passphrase required"))
        || contains(stdErr, QStringLiteral("Passphrase required"))
        || contains(stdOut, QStringLiteral("Encrypted payload"))
        || contains(stdErr, QStringLiteral("Encrypted payload"))
        || contains(stdOut, QStringLiteral("AES decryption failed"))
        || contains(stdErr, QStringLiteral("AES decryption failed"));
}

bool PassphraseManager::isDecryptionFailure(const QString& stdOut, const QString& stdErr) {
    return contains(stdOut, QStringLiteral("AES decryption failed"))
        || contains(stdErr, QStringLiteral("AES decryption failed"))
        || contains(stdOut, QStringLiteral("Encrypted payload"))
        || contains(stdErr, QStringLiteral("Encrypted payload"));
}

QString PassphraseManager::buildPassphraseArg(const QString& passphrase) {
    auto escaped = passphrase;
    escaped.replace('"', QStringLiteral("\\\""));
    return QStringLiteral("--passphrase \"%1\"").arg(escaped);
}
